cartography/graph/loader.py

```diff
@@ -0,0 +1,134 @@
+import logging
+from typing import Dict
+from typing import List
+
+import neo4j
+
+from cartography.graph.querybuilder import build_node_ingestion_query
+from cartography.graph.querybuilder import build_relationship_ingestion_query
+
+logger = logging.getLogger(__name__)
+
+
+def merge_nodes(
+    neo4j_session: neo4j.Session, node_label: str, node_property_map: Dict[str, str],
+    node_data_list: List[Dict], update_tag: int, batch_size: int = 10000,
+) -> None:
+    """
+    Writes the `node_data_list` to Neo4j using the UNWIND pattern in batches with
+    just one MERGE per statement for performance and reliability. Follows guidelines
+    described in
+    - https://neo4j.com/blog/bulk-data-import-neo4j-3-0/
+    - https://dzone.com/articles/tips-for-fast-batch-updates-of-graph-structures-wi
+
+    The query run looks like
+
+    UNWIND {DictList} AS item
+        MERGE (i:`node_label`{id:item.`node_property_map['id']`})
+        ON CREATE SET i.firstseen = timestamp()
+        SET i.lastupdated = {UpdateTag},
+        ... <expand the given node property map to set the other node fields> ...
+
+    :param neo4j_session: The Neo4j session
+    :param node_label: The label of the node to MERGE to Neo4j
+    :param node_property_map: Mapping from node field names to dict keys in `node_data_list`.
+    :param node_data_list: A list of dicts to be written as nodes to the graph.
+    :param update_tag: The cartography update tag.
+    :param batch_size: Optional: the number of nodes to write in a single batch operation.
+    Default = 10000.
+    :return: None.
+    """
+    if batch_size < 1:
+        raise ValueError("batch_size cannot be less than 1.")
+
+    ingest_query = build_node_ingestion_query(node_label, node_property_map)
+
+    chunk = node_data_list[:batch_size]
+    cursor = 0
+    while len(chunk) > 0:
+        query_args = {
+            'UpdateTag': update_tag,
+            'DictList': chunk,
+        }
+        neo4j_session.write_transaction(_merge_chunk, ingest_query, query_args)
```

Comment:
@voutilad I implemented the first step in the spec you commented on - thanks very much for your help and input. Does this usage of `write_transaction()` look correct? I'm wondering if I need to explicitly commit the transaction or anything else. I copied from https://neo4j.com/docs/driver-manual/1.7/sessions-transactions/#driver-transactions-explicit and https://neo4j.com/docs/api/python-driver/1.7/transactions.html.

Reply:
The `write_transaction` method on a `Session` is a managed transaction, so there's no reason to explicitly commit. It will be committed for you and even retried some number of times (3 I think) if there are transient errors like a lock timeout due to a competing transaction.

One thing I'd recommend is explicitly "consuming" the results inside the transaction work function. Technically if the query doesn't have a `RETURN` statement, it's probably ok, but it's a best practice. Not consuming results can lead to transactions being held open. You can either call `consume()` on the result or iterate over the values. (See https://neo4j.com/docs/api/python-driver/current/api.html#result)

Reply:
This is helpful, thanks! I added an update to consume the results. I can't think of a scenario where we would want to return data out of a `merge` query so I'm just having it debuglog counters here.